### PR TITLE
fix: signed division by -1 can overflow

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem\*

Resolves 

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/acir/acir_context/mod.rs [hunk 1]
```diff
@@ -1166,8 +1166,10 @@ impl<F: AcirField> AcirContext<F> {
         let assert_message =
             self.generate_assertion_message_payload("Attempt to divide with overflow".to_string());
         let unsigned = self.not_var(q_sign, AcirType::unsigned(1))?;
-        // We just use `unsigned` for the predicate of assert_neq_var because if the `predicate` is false, the quotient
-        // we get from the unsigned division under the predicate will not be 2^{bit_size-1} anyways.
+
+        // This overflow check must also be under the predicate
+        let unsigned = self.mul_var(unsigned, predicate)?;
```
Comment (asterite) [inline compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: There was a comment here about why we weren't using the predicate var. However, not using it seemed wrong? And using it fixes things so I don't understand the original comment.

Comment (aakoshh) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: This was added only last week https://github.com/noir-lang/noir/pull/9857

Comment (aakoshh) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: This comment looks wrong: https://github.com/noir-lang/noir/blob/31e16d9691ab45206f17ceda0d477ed8d9624d35/compiler/noirc_evaluator/src/acir/acir_context/mod.rs#L1131-L1132 

Shouldn't it have divided by `sign(a)`?

Comment (aakoshh) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: I think what @guipublic meant was that the predicate was used here:
```
let (q1, r1) =
            self.euclidean_division_var(unsigned_lhs, unsigned_rhs, bit_size, predicate)?;
```
So if the `predicate` was `false`, then `q1` will be I suppose zero? And if it's zero, then in this assertion, it will not equal `max_power_of_two`, where `max_power_of_two` is `2^{bit_size-1}`, which will never be 0.
```
        self.assert_neq_var(quotient, max_power_of_two, unsigned, Some(assert_message))?;
```

OTOH I remember we had a discussion that if the `predicate` is off, then witnesses can be assigned any value at all, and although we might assume that the VM will return 0, it might return something else, because it "does not matter" since the predicate is 0 anyway. What if it returned `max_power_of_two` for some reason?

File: compiler/noirc_evaluator/src/ssa/ir/instruction/binary.rs [hunk 2]
```diff
@@ -192,6 +192,22 @@ pub(crate) fn eval_constant_binary_op(
             let Some(rhs) = try_convert_field_element_to_signed_integer(rhs, bit_size) else {
                 return CouldNotEvaluate;
             };
+
+            // Because we always perform signed operations using i128, an operation like `-128_i8 / -1`
+            // will not overflow as it'll actually be done via `-128_i128 / -1`. Thus we need to
+            // manually check this specific case.
+            if matches!(operator, BinaryOp::Div | BinaryOp::Mod) && rhs == -1 {
+                assert!(bit_size < 128);
+                let min_value = -(1_i128 << (bit_size - 1));
```
Comment (aakoshh) [inline compiler/noirc_evaluator/src/ssa/ir/instruction/binary.rs]: NB there is already a `two_pow_bit_size_minus_one` variable further down. I thought maybe it could be reused with some extra comments.

Comment (asterite) [reply compiler/noirc_evaluator/src/ssa/ir/instruction/binary.rs]: Done!

File: tooling/nargo_cli/tests/snapshots/compile_failure/signed_division_by_minus_one_overflow/execute__tests__stderr.snap
Comment (vezenovm) [inline tooling/nargo_cli/tests/snapshots/compile_failure/signed_division_by_minus_one_overflow/execute__tests__stderr.snap:1]: Looks like a leftover snap?

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 3]
```diff
@@ -534,7 +534,29 @@ impl Instruction {
...
+                    }
+
+                    // For signed types, division or modulo by -1 can overflow.
+                    let NumericType::Signed { bit_size } =
+                        dfg.type_of_value(binary.rhs).unwrap_numeric()
+                    else {
+                        return false;
+                    };
+
+                    let minus_one_as_field = FieldElement::from((1_u128 << bit_size) - 1);
```
Comment (vezenovm) [inline compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Could we reuse `ssa::ir::integer` here?

File: compiler/noirc_evaluator/src/acir/acir_context/mod.rs (hunk 1 above)
Comment (vezenovm) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: Yeah the comment refers to the actual `euclidean_division_var` that gives the quotient. I guess we may sometimes have a case where the quotient is `2^{bit_size-1}` under a false predicate 🤔. Any thoughts here @guipublic? 

@asterite Could you try just using the `predicate` directly in `assert_neq_var` rather than `unsigned`.

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs (hunk 3 above)
Comment (asterite) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Sorry, how?

Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Could we not convert `rhs` with its type to a `IntegerConstant` then compare it against `IntegerConstant::Signed { value: -1, bit_size }`?

File: compiler/noirc_evaluator/src/acir/acir_context/mod.rs (hunk 1 above)
Comment (asterite) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: I can try it... but wouldn't it be incorrect to not check for unsigned? Otherwise I think if it's a signed value and the underlying field element matches `2^{bit_size-1}` then it will overflow? Not sure...

Comment (asterite) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: I just tried it. The program `regression_8726` fails with that change.

Comment (vezenovm) [reply compiler/noirc_evaluator/src/acir/acir_context/mod.rs:1171]: Ah I'm dumb. I was thinking cause we are in `signed_division_var` it was fine, but we do still need to check the sign.

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs (hunk 3 above)
Comment (asterite) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs]: Nice! I didn't know about this type. Done.

**Comment by asterite** [comment]
> Looks good. Before we merge can you check if we solve this too https://github.com/noir-lang/noir/issues/10008

Yes, it solves that too.

Should I add a regression test for it? I'm not sure because maybe in the end it's a duplicate of some of the other issues...

**Comment by vezenovm** [comment]
> Should I add a regression test for it? I'm not sure because maybe in the end it's a duplicate of some of the other issues...

That test has a dynamic divisor. So if you see a test with a dynamic divisor no need, but otherwise doesn't hurt.
